apps/rush-lib/src/api/LastInstallFlag.ts

```diff
@@ -3,111 +3,126 @@
  * It also compares state, so that if something like the Node.js version has changed,
  * it can invalidate the last install.
  * @internal
  */
-export class LastInstallFlag {
-  private _path: string;
-  private _state: JsonObject;
-
+export class LastInstallFlag extends FlagFileBase<ILastInstallFlagJson> {
   /**
    * Creates a new LastInstall flag
    * @param folderPath - the folder that this flag is managing
    * @param state - optional, the state that should be managed or compared
    */
-  public constructor(folderPath: string, state: JsonObject = {}) {
-    this._path = path.join(folderPath, LAST_INSTALL_FLAG_FILE_NAME);
-    this._state = state;
+  public constructor(folderPath: string, state: ILastInstallFlagJson = {}) {
+    super(path.join(folderPath, LAST_INSTALL_FLAG_FILE_NAME), state);
   }
 
   /**
-   * Returns true if the file exists and the contents match the current state.
+   * Gets the current state of the Rush repo. This state is used to compare against
+   * the last-known-good state tracked by the LastInstall flag.
+   * @param rushConfiguration - the configuration of the Rush repo to get the install
+   * state from
    */
-  public isValid(): boolean {
-    return this._isValid(false);
+  public static getCurrentState(rushConfiguration: RushConfiguration): ILastInstallFlagJson {
+    const currentState: ILastInstallFlagJson = {
+      node: process.versions.node,
+      packageManager: rushConfiguration.packageManager,
+      packageManagerVersion: rushConfiguration.packageManagerToolVersion
+    };
+
+    if (currentState.packageManager === 'pnpm' && rushConfiguration.pnpmOptions) {
+      currentState.storePath = rushConfiguration.pnpmOptions.pnpmStorePath;
+      if (rushConfiguration.pnpmOptions.useWorkspaces) {
+        currentState.workspaces = rushConfiguration.pnpmOptions.useWorkspaces;
+      }
+    }
+
+    return currentState;
   }
 
   /**
-   * Same as isValid(), but with an additional check:  If the current state is not equal to the previous
-   * state, and an the current state causes an error, then throw an exception with a friendly message.
+   * Returns true if the file exists and the contents match the current state.
+   * @param reportStoreIssues - default false, validation will throw if the package manager store
+   * is not valid.
    */
-  public checkValidAndReportStoreIssues(): boolean {
-    return this._isValid(true);
-  }
-
-  private _isValid(checkValidAndReportStoreIssues: boolean): boolean {
-    if (!FileSystem.exists(this._path)) {
-      return false;
-    }
-
-    let oldState: JsonObject;
-    try {
-      oldState = JsonFile.load(this._path);
-    } catch (err) {
+  public isValid(reportStoreIssues: boolean = false): boolean {
+    const oldState: ILastInstallFlagJson | undefined = this.loadFromFile();
+    if (!oldState) {
       return false;
+    } else if (_.isEqual(oldState, this.state)) {
+      return true;
     }
```

Comment:
Should this just be in the base?

Reply:
This was the original design I had. However, if a class wanted to override and perform additional validation on the state that was loaded from file (such as here with the store path), they would need to load the state from file a second time, which is not ideal. Abstracting additional validation into a separate abstract method was also a consideration, but then there's no good way to forward on the `reportStoreIssues` param. So at the end of the day, I just left validation completely up to the implementer. Let me know if you have any better ideas around this :)